packager/packager.cc

```diff
@@ -0,0 +1,665 @@
+    vector<shared_ptr<PackageInfo>> packages;
+    {
+        Timer timeit(gs.tracer(), "packager.findPackages");
+        core::UnfreezeNameTable unfreeze(gs);
+        for (auto &file : files) {
+            if (FileOps::getFileName(file.file.data(gs).path()) == PACKAGE_FILE_NAME) {
+                file.file.data(gs).sourceType = core::File::Type::Package;
+                core::MutableContext ctx(gs, core::Symbols::root(), file.file);
+                auto pkg = getPackageInfo(ctx, file);
+                // getPackageInfo emits an error if the package is malformed.
+                if (pkg != nullptr) {
+                    auto &existing = packageInfoByMangledName[pkg->name.mangledName];
+                    if (existing != nullptr) {
+                        if (auto e =
+                                ctx.beginError(pkg->loc.offsets(), core::errors::Packager::RedefinitionOfPackage)) {
+                            auto pkgName = absl::StrJoin(pkg->name.fullName, "::", NameFormatter(ctx));
+                            e.setHeader("Redefinition of package `{}`", pkgName);
+                            e.addErrorLine(existing->loc, "Package `{}` originally defined here", pkgName);
+                        }
+                    } else {
+                        existing = pkg;
+                    }
+                    packageInfoByFile[file.file] = pkg;
+                    packages.emplace_back(move(pkg));
+                }
+            }
+        }
+    }
+
+    // Sort packages so that packages with the longest/most specific paths are first.
+    // That way, the first package to match a file's path is the most specific package match.
+    fast_sort(packages, [](const auto &a, const auto &b) -> bool {
+        return a->packagePathPrefix.size() > b->packagePathPrefix.size();
+    });
+
+    {
+        Timer timeit(gs.tracer(), "packager.rewritePackages");
+        // Step 2: Rewrite packages. Can be done in parallel (and w/ step 3) if this becomes a bottleneck.
+        for (auto &file : files) {
+            if (file.file.data(gs).sourceType == core::File::Type::Package) {
+                auto &packageInfo = packageInfoByFile[file.file];
+                if (packageInfo != nullptr) {
+                    core::Context ctx(gs, core::Symbols::root(), file.file);
+                    file = rewritePackage(ctx, move(file), *packageInfo, packageInfoByMangledName);
+                }
+            }
+        }
+    }
+
+    // Step 3: Find files within each package and rewrite each.
+    {
+        Timer timeit(gs.tracer(), "packager.rewritePackagedFiles");
+
+        auto resultq = make_shared<BlockingBoundedQueue<vector<ast::ParsedFile>>>(files.size());
+        auto fileq = make_shared<ConcurrentBoundedQueue<ast::ParsedFile>>(files.size());
+        for (auto &file : files) {
+            fileq->push(move(file), 1);
+        }
+
+        workers.multiplexJob("rewritePackagedFiles", [&gs, &packages, fileq, resultq]() {
```

Comment:
I miss the days where all the multiplexing was contained to `pipeline.cc` 😞

Reply:
Sorry, I didn't realize that was a constraint. :(

I could try to lift the logic there so `pipeline.cc` is driving the process more?

Reply:
No, we've already broken this by adding parallelism to both resolver and namer. I'm just noting that we're seeing it in one more place.